### ISSUE TITLE
[Roadmap] Option to customize local LLM inference settings

### ISSUE THREAD

**Issue by Daniel-Dan-Espinoza** [enricoros/big-AGI#343]
**Why**
Would provide users with more control over the inference settings such as  "top_k", "min_p", "top_p", "mirostat", etc.

**Concise description**
An advanced settings menu where the requests sent to the local LLM can be customized. Either as a slider menu where the inference settings such as  "top_k", "min_p", "top_p", "mirostat", etc., can be selected or a textbox where inference configuration can be entered.

**Requirements**
Additional settings to customize local LLM model settings in 'Configure AI Models' menu. Option to import the inference configuration from a JSON file.

**Comment by enricoros** [comment]
Hi @Daniel-Dan-Espinoza , right now we have the model options dialog on each model, that looks something like this:
![image](https://github.com/enricoros/big-AGI/assets/32999/21fcdf8e-04e3-4239-921e-d33afaaded15)

Are you asking for a different mechanics/dialog, or more options on this dialog?
Of the options (min_p, top_k, top_p, ..) could you prioritize which ones you use the most? We have 10 providers (11 soon) and need to add/test each of their models, so we need prioritization. What's mirostat by the way?

**Comment by Daniel-Dan-Espinoza** [comment]
> Hi @Daniel-Dan-Espinoza , right now we have the model options dialog on each model, that looks something like this: ![image](https://private-user-images.githubusercontent.com/32999/296867352-21fcdf8e-04e3-4239-921e-d33afaaded15.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUzNzY4MDEsIm5iZiI6MTcwNTM3NjUwMSwicGF0aCI6Ii8zMjk5OS8yOTY4NjczNTItMjFmY2RmOGUtMDRlMy00MjM5LTkyMWUtZDMzYWZhYWRlZDE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE2VDAzNDE0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzYmUzOTQ5OTY1MTE3MWE3ZDg5ZGU3MGQ4Y2MyM2MxMDMzNjA5ZTg4ZDJkMmRiODFlYmNkODYwM2Y0NDBhNTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.laQQkcsehh6m8nvRjR1el8B0jNtWUrZ85NTaoNV1gO4)
> 
> Are you asking for a different mechanics/dialog, or more options on this dialog? Of the options (min_p, top_k, top_p, ..) could you prioritize which ones you use the most? We have 10 providers (11 soon) and need to add/test each of their models, so we need prioritization. What's mirostat by the way?

I would prefer adding more options on the above dialog box. 

These are hyperparameters that control the inference in LLMs run locally using Koboldcpp, LM Studios, etc. Refer https://github.com/ggerganov/llama.cpp/blob/master/examples/main/README.md#generation-flags.

**Top-K:** This setting limits the number of possible words to choose from to the top K most likely options, removing everything else. Can be used with Top-P. Set value to 0 to disable its effect.
*"top_k"*: A higher value for top-k (e.g., 100) will consider more tokens and lead to more diverse text, while a lower value (e.g., 10) will focus on the most probable tokens and generate more conservative text (default:40).

**Top-P:** Discards unlikely text in the sampling process. Only considers words with the highest cumulative probabilities summing up to P. Low values make the text predictable, as uncommon tokens are removed. Set value to 1 to disable its effect.
*"top_p"*: A higher value for top-p (e.g., 0.95) will lead to more diverse text, while a lower value (e.g., 0.5) will generate more focused and conservative text (default:0.9).

**Mirostat Sampling**: A newer sampling method that adjusts the value of k in top-k decoding to keep the perplexity within a specific range. It prevents the generated text from becoming repetitive, and losing coherence. Refer (https://github.com/ggerganov/llama.cpp/blob/master/examples/main/README.md#mirostat-sampling). Requires 3 parameters.
*"mirostat"*: Enable Mirostat sampling, controlling perplexity during text generation (default: 0, 0 = disabled, 1 = Mirostat, 2 = Mirostat 2.0).
*"mirostat_tau"*: Set the Mirostat target entropy, parameter tau, which represents the desired perplexity value for the generated text. Adjusting the target entropy allows you to control the balance between coherence and diversity in the generated text. A lower value will result in more focused and coherent text, while a higher value will lead to more diverse and potentially less coherent text (default: 5.0).
*"mirostat_eta"*: Set the Mirostat learning rate, parameter eta. Influences how quickly the algorithm responds to feedback from the generated text. A lower learning rate will result in slower adjustments, while a higher learning rate will make the algorithm more responsive (default: 0.1).

From what I know, the parameters are passed to the backend in the below form along with the prompt. I do not know the method you use to pass it through the api, but I have seen these parameters passed to the LM Studio in the Server Logs while using big-AGI. Even if it is not possible to add sliders for these settings in the dialog box, any option to edit the below inference configuration being passed to the LM Studio would be fine. 

LM Studio Server logs:
```
[2024-01-16 10:33:48.530] [INFO] [LM STUDIO SERVER] Processing queued request...
[2024-01-16 10:33:48.531] [INFO] Received POST request to /v1/chat/completions with body: {
  "model": "D:\\LLMs\\Phi2Dolphin\\dolphin-2_6-phi-2.gguf",
  "messages": [
    {
      "role": "system",
      "content": "You are ChatGPT, a large language model trained by OpenAI, based on the GPT-4 architecture.\nKnowledge cutoff: 2021-09\nCurrent date: 2024-01-16\n"
    },
    {
      "role": "user",
      "content": "This is a test"
    }
  ],
  "temperature": 0.7,
  "stream": true
}
[2024-01-16 10:33:48.532] [INFO] [LM STUDIO SERVER] Context Overflow Policy is: Stop at limit
[2024-01-16 10:33:48.533] [INFO] Provided inference configuration: {
  "n_threads": 1,
  "n_predict": -1,
  "top_k": 40,
  "min_p": 0.05,
  "top_p": 0.95,
  "temp": 0.7,
  "repeat_penalty": 1.1,
  "input_prefix": "<|im_end|>\\n<|im_start|>user\\n",
  "input_suffix": "<|im_end|>\\n<|im_start|>assistant\\n",
  "antiprompt": [
    "<|im_start|>",
    "<|im_end|>"
  ],
  "pre_prompt": "You are ChatGPT, a large language model trained by OpenAI, based on the GPT-4 architecture.\nKnowledge cutoff: 2021-09\nCurrent date: 2024-01-16\n",
  "pre_prompt_suffix": "",
  "pre_prompt_prefix": "<|im_start|>system\\n",
  "seed": -1,
  "tfs_z": 1,
  "typical_p": 1,
  "repeat_last_n": 64,
  "frequency_penalty": 0,
  "presence_penalty": 0,
  "n_keep": 0,
  "logit_bias": {},
  "mirostat": 0,
  "mirostat_tau": 5,
  "mirostat_eta": 0.1,
  "memory_f16": true,
  "multiline_input": false,
  "penalize_nl": true
}
```

**Comment by enricoros** [comment]
The per-model options dialog is now backed by a typed parameter system, and top_p is in it - top_k, min_p and the mirostat trio never were. Rescoping this to exactly that: sampling parameters declared per model and wired for the local vendors (Ollama, LM Studio, LocalAI). Thanks for the writeup, it is still the spec for this.
